### ISSUE TITLE
Add comment formatting via a feature call

### ISSUE THREAD

**Issue by jdanish** [theRAPTLab/gsgo#758]
Right now the different comment styles are handled via code that requires a re-build. Ideally we'd have some defaults but be able to add others from a feature call so it can happen all in a gem-step project. 

Bonus if we can see the options similar to the enum function.

This is also a bottom level of priority for this time period.

**Comment by benloh** [comment]
Given the three issues...
* #752 
* #751 
* #758
...we probably want to consider implementing all of this as a new comment keyword that accepts a type and text?


A few complications:
* multiple lines -- currently script lines are only allowed on a single line.  They do not wrap.  Would line numbering be messed up if we allow long comment lines to wrap?  Or does the js rendering handle that?
* enumerated types -- given #758 we need to dynamically define types.

If we're implementing enums, then we may as well do that first.

**Comment by benloh** [comment]
@jdanish Currently you essentially have five different comment styles:
1. explanationCommentHeader
2. explanationCommentBody
3. changeCommentHeader
4. changeCommentBody
5. commentKeyHeader

If we introduce project-customizable comment styles, would your comments map to one of the existing types (relatively simple)?  Or are you wanting the project settings to be able to define arbitrary styles (can get really hairy if we're defining custom color, bgcolor, and font-style via the project setting UI, and EVEN messier if project settings are defining css)

**Comment by jdanish** [comment]
I think ideally we’d want something defined by the project. But we could agree to limit it to foreground and background color if that’d help?

If we are going to map to styles I’d want to create a few more to be safe and then let’s see how it goes. Would that work?

Joshua

**Comment by benloh** [comment]
A few more questions:
1. Who will be making these comment style changes?  Just researchers?  Or also teachers?  Students?
2. How often would you make these changes?  Just once in a blue moon?
3. Will customized styles be shared with across a lot of projects?
4. Is it important to have a UI (e.g. project setting edit UI) for making those changes?  Or can we just use a TOML settings file?  For example, we can just add a generic `_commentstyles.toml` file to `art-assets` (or perhaps in the main codebase if it should be common toall) for specifying a common set of comment styles (the ones you've already defined), and then add a project-specific styles, e.g. `wetlands_beavers_story.styles.toml` file to add project-specific customizations?

**Comment by jdanish** [comment]
I’ve lost track of which approach you are taking, so to make sure I am getting this right, the idea is that there would be:

1) Some named CSS style someplace
2) Only researchers would edit those
3) The list of names will somehow be available to kids / teachers
4) The kids / teachers can reference them

If that is the case, then I think we can have one set of styles per install. 

If you are saying it could be done via the script, like in global, then likely we’d want some default styles and override them by project. I imagine that’s a lot more work though, so am OK with the approach above.

**Comment by jdanish** [comment]
In reference to your summary, yes that sounds right.

Thinking this over now that I see all the details, I think let's just assuming a set of common styles for the entire install, so IU has a set of styles for all projects. Period.

Ideally it won't require a recompile but if it does we can even work around that for the current funding. I think we'd rather put all the thinking about complexity elsewhere.

**Comment by benloh** [comment]
Thanks.  

Just to make sure I understand.  "IU has a set of styles for all projects" -- so VU shares the same set?  Or does VU have it's own set?

And, just checking assumptions about "recompiling"...currently the comment style definitions have been moved out of `SharedElements.tsx` and put into a single file `common-utilities.ts`.  I will probably also move the css style classes out of `gem-ui.css` and either figure out how to inject those classes into `common-utilities.ts` or via some other mechanism so that all the styles are programmatically defined in a single place.  It should be much more easy to change for someone like you or Morgan.  But it is not quite user-friendly for a non-technical user.

e.g. the current definitions look something like this: `comment-utilties.ts:47`
```
COMMENTTYPEMAP.set('🔎 QUESTION', {
  style: 'explanationCommentHeader',
  isBookmark: false
});
COMMENTTYPEMAP.set('✏️ LETS', {
  style: 'changeCommentHeader',
  isBookmark: true
});
```
You can also imagine we might migrate to something like this:
```
COMMENTTYPEMAP.set('🔎 QUESTION', {
  color: '#ed8b01',
  isReversed: true,    // for header white text on colored background
  isBookmark: false
});
COMMENTTYPEMAP.set('🔎', {
  color: '#43a873',
  isReversed: false     // for body colored text on white background
  isBookmark: true
});
```
...so you'd still have to modify code to set the styles, but now you can easily create as many arbitrary styles as you want, and do it all in one place.  But you still have to recompile (e.g. run `npm run gem`) to update the changes.  Is that sufficient?  I'm not sure I'm quite understanding the need to avoid recompiling, unless you're thinking we are electron-ifying the app?

**Comment by benloh** [comment]
Got it.  Thanks.  So that does suggest that we want to introduce a common `_commentstyle.toml` file to `art-assets` so that you can define styles for the whole `art-assets` repo and to do so in a way that allows diferent sets.

How are you currently handling `art-assets` for IU vs VU?  Are they on different branches?  different folders?  Do we need a more sophisticated way of handling the different projects?

**Comment by jdanish** [comment]
Right now we use different branches if / when we need to, but we generally can use the same branch since we are actually working in different models currently.

We do not use different folders. It is easier for folks to use a different branch than to find and change the local settings file to switch folders even though both should be easy.

I think using git or moving folders by hand is fine by now, so while it might be nice to have sub-folders and other things long-term, this all works great for now, and putting a single comment style file in art-assets sounds good.

**Comment by benloh** [comment]
Implemented with #763
